// @gov.au/pancake-js/lib/pancake.js
/***************************************************************************************************************************************************************
 *
 * Plug-in for Pancake
 *
 * Move and uglify js files from pancake modules into your pancake folder
 *
 * @repo    - https://github.com/govau/pancake
 * @author  - Dominik Wilkowski
 * @license - https://raw.githubusercontent.com/govau/pancake/master/LICENSE (MIT)
 *
 **************************************************************************************************************************************************************/
'use strict'; //--------------------------------------------------------------------------------------------------------------------------------------------------------------
// Dependencies
//--------------------------------------------------------------------------------------------------------------------------------------------------------------

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

var Path = require('path');

var Fs = require('fs'); //--------------------------------------------------------------------------------------------------------------------------------------------------------------
// Module imports
//--------------------------------------------------------------------------------------------------------------------------------------------------------------


var _require = require('@gov.au/pancake'),
    Log = _require.Log,
    Style = _require.Style,
    Loading = _require.Loading,
    ReadFile = _require.ReadFile,
    WriteFile = _require.WriteFile;

var _require2 = require('./js'),
    HandleJS = _require2.HandleJS,
    MinifyAllJS = _require2.MinifyAllJS;

Log.output = true; //this plugin assumes you run it through pancake
//--------------------------------------------------------------------------------------------------------------------------------------------------------------
// Plugin export
//--------------------------------------------------------------------------------------------------------------------------------------------------------------

/**
 * The main pancake method for this plugin
 *
 * @param  {array}  version        - The version of mother pancake
 * @param  {array}  modules        - An array of all module objects
 * @param  {object} settings       - An object of the host package.json file and it’s path
 * @param  {object} GlobalSettings - An object of the global settings
 * @param  {object} cwd            - The path to the working directory of our host package.json file
 *
 * @return {Promise object}  - Returns an object of the settings we want to save
 */

module.exports.pancake = function (version, modules, settings, GlobalSettings, cwd) {
  Loading.start('pancake-js', Log.verboseMode); //--------------------------------------------------------------------------------------------------------------------------------------------------------------
  // Settings
  //--------------------------------------------------------------------------------------------------------------------------------------------------------------

  var SETTINGS = {
    js: {
      minified: true,
      modules: false,
      location: 'pancake/js/',
      name: 'pancake.min.js'
    }
  }; //merging settings with host settings

  Object.assign(SETTINGS.js, settings.js);
  return new Promise(function (resolve, reject) {
    //some housekeeping
    if (typeof version !== 'string') {
      reject("Plugin pancake-js got a mismatch for the data that was passed to it! ".concat(Style.yellow("version"), " was ").concat(Style.yellow(_typeof(version)), " ") + "but should have been ".concat(Style.yellow("string")));
    }

    if (_typeof(modules) !== 'object') {
      reject("Plugin pancake-js got a mismatch for the data that was passed to it! ".concat(Style.yellow("modules"), " was ").concat(Style.yellow(_typeof(modules)), " ") + "but should have been ".concat(Style.yellow("object")));
    }

    if (_typeof(settings) !== 'object') {
      reject("Plugin pancake-js got a mismatch for the data that was passed to it! ".concat(Style.yellow("settings"), " was ").concat(Style.yellow(_typeof(settings)), " ") + "but should have been ".concat(Style.yellow("object")));
    }

    if (typeof cwd !== 'string') {
      reject("Plugin pancake-js got a mismatch for the data that was passed to it! ".concat(Style.yellow("cwd"), " was ").concat(Style.yellow(_typeof(cwd)), " ") + "but should have been ".concat(Style.yellow("string")));
    } //--------------------------------------------------------------------------------------------------------------------------------------------------------------
    // Settings
    //--------------------------------------------------------------------------------------------------------------------------------------------------------------


    var compiledAll = []; //for collect all promises
    //--------------------------------------------------------------------------------------------------------------------------------------------------------------
    // Iterate over each module
    //--------------------------------------------------------------------------------------------------------------------------------------------------------------

    var _iteratorNormalCompletion = true;
    var _didIteratorError = false;
    var _iteratorError = undefined;

    try {
      for (var _iterator = modules[Symbol.iterator](), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true) {
        var modulePackage = _step.value;
        Log.verbose("JS: Building ".concat(Style.yellow(modulePackage.name))); //check if there are js files

        var jsModulePath = void 0;

        if (modulePackage.pancake['pancake-module'].js !== undefined) {
          jsModulePath = Path.normalize("".concat(modulePackage.path, "/").concat(modulePackage.pancake['pancake-module'].js.path));
        }

        if (!Fs.existsSync(jsModulePath)) {
          Log.verbose("JS: No js found in ".concat(Style.yellow(jsModulePath)));
        } else {
          Log.verbose("JS: ".concat(Style.green('⌘'), " Found Javascript files in ").concat(Style.yellow(jsModulePath)));
          var jsModuleToPath = Path.normalize("".concat(cwd, "/").concat(SETTINGS.js.location, "/").concat(modulePackage.name.split('/')[1], ".js")); //compile js and write to file depending on settings

          var jsPromise = HandleJS(jsModulePath, SETTINGS.js, jsModuleToPath, "".concat(modulePackage.name, " v").concat(modulePackage.version)).catch(function (error) {
            Log.error(error);
          });
          compiledAll.push(jsPromise); //collect all js promises so we can save the SETTINGS.js.name file later
        }
      }
    } catch (err) {
      _didIteratorError = true;
      _iteratorError = err;
    } finally {
      try {
        if (!_iteratorNormalCompletion && _iterator.return != null) {
          _iterator.return();
        }
      } finally {
        if (_didIteratorError) {
          throw _iteratorError;
        }
      }
    }

    if (modules.length < 1) {
      Loading.stop('pancake-js', Log.verboseMode); //stop loading animation

      Log.info("No pancake modules found \uD83D\uDE2C");
      resolve(SETTINGS);
    } else {
      //write SETTINGS.js.name file
      if (SETTINGS.js.name !== false) {
        compiledAll.push(MinifyAllJS(version, compiledAll, SETTINGS.js, cwd).catch(function (error) {
          Log.error(error);
        }));
      } //after all files have been compiled and written


      Promise.all(compiledAll).catch(function (error) {
        Loading.stop('pancake-js', Log.verboseMode); //stop loading animation

        Log.error("Js plugin ran into an error: ".concat(error));
      }).then(function () {
        Log.ok('JS PLUGIN FINISHED');
        Loading.stop('pancake-js', Log.verboseMode); //stop loading animation

        resolve(SETTINGS);
      });
    }
  });
};
